File: src/algorithms/pathfinding/maze-generation/recursive-division/sources/recursive-division_test.go
package recursivedivision

import "testing"

func makeOpenGrid(rows, cols int) [][]GridCell {
	grid := make([][]GridCell, rows)
	for row := range grid {
		grid[row] = make([]GridCell, cols)
		for col := range grid[row] {
			grid[row][col] = GridCell{Row: row, Col: col, CellType: CellEmpty, State: "default"}
		}
	}
	return grid
}

func bfsReachable(grid [][]GridCell, startRow, startCol, endRow, endCol int) bool {
	rowCount, colCount := len(grid), len(grid[0])
	visited := make([][]bool, rowCount)
	for row := range visited {
		visited[row] = make([]bool, colCount)
	}
	type pos struct{ row, col int }
	queue := []pos{{startRow, startCol}}
	visited[startRow][startCol] = true
	dirs := []pos{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}
	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]
		if curr.row == endRow && curr.col == endCol {
			return true
		}
		for _, dir := range dirs {
			nr, nc := curr.row+dir.row, curr.col+dir.col
			if nr >= 0 && nr < rowCount && nc >= 0 && nc < colCount && !visited[nr][nc] && grid[nr][nc].CellType != CellWall {
				visited[nr][nc] = true
				queue = append(queue, pos{nr, nc})
			}
		}
	}
	return false
}

func TestBuildsWalls(t *testing.T) {
	grid := makeOpenGrid(9, 9)
	grid[1][1].CellType = CellStart
	grid[7][7].CellType = CellEnd
	result := RecursiveDivision(grid, 1, 1, 7, 7)
	if result.WallsBuilt == 0 {
		t.Error("expected wallsBuilt > 0")
	}
}

func TestStartAndEndPreserved(t *testing.T) {
	grid := makeOpenGrid(9, 9)
	grid[1][1].CellType = CellStart
	grid[7][7].CellType = CellEnd
	RecursiveDivision(grid, 1, 1, 7, 7)
	if grid[1][1].CellType != CellStart {
		t.Error("start cell should be preserved")
	}
	if grid[7][7].CellType != CellEnd {
		t.Error("end cell should be preserved")
	}
}

func TestPathStillExists(t *testing.T) {
	grid := makeOpenGrid(9, 9)
	grid[1][1].CellType = CellStart
	grid[7][7].CellType = CellEnd
	RecursiveDivision(grid, 1, 1, 7, 7)
	if !bfsReachable(grid, 1, 1, 7, 7) {
		t.Error("start should still reach end after division")
	}
}

func TestWallsActuallyAdded(t *testing.T) {
	grid := makeOpenGrid(9, 9)
	grid[1][1].CellType = CellStart
	grid[7][7].CellType = CellEnd
	RecursiveDivision(grid, 1, 1, 7, 7)
	wallCount := 0
	for _, row := range grid {
		for _, cell := range row {
			if cell.CellType == CellWall {
				wallCount++
			}
		}
	}
	if wallCount == 0 {
		t.Error("expected wall cells in grid after division")
	}
}

File: src/algorithms/pathfinding/maze-generation/recursive-division/__tests__/RecursiveDivision_test.cpp
#include "../sources/RecursiveDivision.cpp"
#include <cassert>
#include <iostream>
#include <queue>

std::vector<std::vector<GridCell>> makeOpenGrid(int rows, int cols) {
  std::vector<std::vector<GridCell>> grid(rows, std::vector<GridCell>(cols));
  for (int row = 0; row < rows; row++)
    for (int col = 0; col < cols; col++)
      grid[row][col] = {row, col, CellType::Empty, "default"};
  return grid;
}

bool bfsReachable(const std::vector<std::vector<GridCell>>& grid, int startRow, int startCol, int endRow, int endCol) {
  int rows = (int)grid.size(), cols = (int)grid[0].size();
  std::vector<std::vector<bool>> visited(rows, std::vector<bool>(cols, false));
  std::queue<std::pair<int,int>> bfsQueue;
  bfsQueue.push({startRow, startCol});
  visited[startRow][startCol] = true;
  int deltaRows[] = {-1, 1, 0, 0}, deltaCols[] = {0, 0, -1, 1};
  while (!bfsQueue.empty()) {
    auto [row, col] = bfsQueue.front(); bfsQueue.pop();
    if (row == endRow && col == endCol) return true;
    for (int dir = 0; dir < 4; dir++) {
      int nextRow = row + deltaRows[dir], nextCol = col + deltaCols[dir];
      if (nextRow >= 0 && nextRow < rows && nextCol >= 0 && nextCol < cols
          && !visited[nextRow][nextCol] && grid[nextRow][nextCol].cellType != CellType::Wall) {
        visited[nextRow][nextCol] = true;
        bfsQueue.push({nextRow, nextCol});
      }
    }
  }
  return false;
}

int main() {
  // Test: builds walls
  {
    auto grid = makeOpenGrid(9, 9);
    grid[1][1].cellType = CellType::Start;
    grid[7][7].cellType = CellType::End;
    auto result = recursiveDivision(grid, {1, 1}, {7, 7});
    assert(result.wallsBuilt > 0);
  }

  // Test: start and end preserved
  {
    auto grid = makeOpenGrid(9, 9);
    grid[1][1].cellType = CellType::Start;
    grid[7][7].cellType = CellType::End;
    recursiveDivision(grid, {1, 1}, {7, 7});
    assert(grid[1][1].cellType == CellType::Start);
    assert(grid[7][7].cellType == CellType::End);
  }

  // Test: path still exists after division
  {
    auto grid = makeOpenGrid(9, 9);
    grid[1][1].cellType = CellType::Start;
    grid[7][7].cellType = CellType::End;
    recursiveDivision(grid, {1, 1}, {7, 7});
    assert(bfsReachable(grid, 1, 1, 7, 7));
  }

  // Test: walls actually added to grid
  {
    auto grid = makeOpenGrid(9, 9);
    grid[1][1].cellType = CellType::Start;
    grid[7][7].cellType = CellType::End;
    recursiveDivision(grid, {1, 1}, {7, 7});
    int wallCount = 0;
    for (const auto& row : grid)
      for (const auto& cell : row)
        if (cell.cellType == CellType::Wall) wallCount++;
    assert(wallCount > 0);
  }

  std::cout << "All tests passed!" << std::endl;
  return 0;
}
